core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/JDBCSourceAsColumnarStore.scala

```diff
@@ -65,13 +66,121 @@ class JDBCSourceAsColumnarStore(override val connProperties: ConnectionPropertie
+      (conn: Connection) => {
+        connectionType match {
+          case ConnectionType.Embedded =>
+            val txMgr = Misc.getGemFireCache().getCacheTransactionManager()
+            if (txMgr.getTXState() == null) {
+              txMgr.begin(com.gemstone.gemfire.cache.IsolationLevel.SNAPSHOT, null)
+              Array(conn, txMgr.getTransactionId().stringFormat())
+            } else {
+              Array(conn, null)
+            }
+          case _ =>
+            val txId = SparkShellRDDHelper.snapshotTxIdForWrite.get
+            if (txId == null) {
+              logDebug("Going to start the transaction on server ")
+              val startAndGetSnapshotTXId = conn.prepareCall(s"call sys.START_SNAPSHOT_TXID (?)")
+              startAndGetSnapshotTXId.registerOutParameter(1, java.sql.Types.VARCHAR)
+              startAndGetSnapshotTXId.execute()
+              val txid: String = startAndGetSnapshotTXId.getString(1)
+              startAndGetSnapshotTXId.close()
+              SparkShellRDDHelper.snapshotTxIdForWrite.set(txid)
+              logDebug(s"The snapshot tx id is ${txid} and tablename is ${tableName}")
+              Array(conn, txid)
+            } else {
+              // it should always be not null.
+              if (!txId.equals("null")) {
+                val statement = conn.createStatement()
+                statement.execute(
+                  s"call sys.USE_SNAPSHOT_TXID('$txId')")
+              }
+              Array(conn, null)
+            }
+        }
+      }
+    } (implicitly, Some(conn))
+  }
+
+  def commitTx(txId: String)(implicit conn: Option[Connection]): Unit = {
+    tryExecute(tableName, closeOnSuccessOrFailure = true, onExecutor = true) {
+      (conn: Connection) => {
+        connectionType match {
+          case ConnectionType.Embedded =>
+            Misc.getGemFireCache().getCacheTransactionManager().commit()
+          case _ =>
+            logDebug(s"Going to commit ${txId} the transaction on server ")
+            val ps = conn.prepareStatement(s"call sys.COMMIT_SNAPSHOT_TXID(?)")
+            ps.setString(1, if (txId == null) "null" else txId)
+            ps.executeUpdate()
+            logDebug(s"The txid being committed is $txId")
+            ps.close()
+            SparkShellRDDHelper.snapshotTxIdForWrite.set(null)
+            logDebug(s"Committed ${txId} the transaction on server ")
+        }
+      }
+    }(implicitly, conn)
+  }
+
+
+  def rollbackTx(txId: String)(implicit conn: Option[Connection]): Unit = {
+    tryExecute(tableName, closeOnSuccessOrFailure = true, onExecutor = true) {
+      (conn: Connection) => {
+        connectionType match {
+          case ConnectionType.Embedded =>
+            Misc.getGemFireCache().getCacheTransactionManager().rollback()
+          case _ =>
+            logDebug(s"Going to rollback ${txId} the transaction on server ")
+            val ps = conn.prepareStatement(s"call sys.ROLLBACK_SNAPSHOT_TXID(?)")
```

Comment:
Is there a way we can avoid this preparation everytime for rollback, commits, fetch etc. Ideally for one connection we just need this to happen once. Isn't it? Please think about this. If easily doable do it now but make a note of it. May be file a JIRA task if you thnk this should be done.